internal/librarian/stateandconfig.go

```diff
@@ -87,17 +91,74 @@ func parsePipelineConfig(contentLoader func() ([]byte, error)) (*config.Pipeline
 	return config, nil
 }
 
-func parseLibrarianState(contentLoader func() ([]byte, error)) (*config.LibrarianState, error) {
-	bytes, err := contentLoader()
+func parseLibrarianState(contentLoader func(file string) ([]byte, error), path, source string) (*config.LibrarianState, error) {
+	bytes, err := contentLoader(path)
 	if err != nil {
 		return nil, err
 	}
 	var s config.LibrarianState
 	if err := yaml.Unmarshal(bytes, &s); err != nil {
 		return nil, fmt.Errorf("unmarshaling librarian state: %w", err)
 	}
+	if err := populateServiceConfig(&s, contentLoader, source); err != nil {
+		return nil, fmt.Errorf("populating service config: %w", err)
+	}
 	if err := s.Validate(); err != nil {
 		return nil, fmt.Errorf("validating librarian state: %w", err)
 	}
 	return &s, nil
 }
+
+func populateServiceConfig(state *config.LibrarianState, contentLoader func(file string) ([]byte, error), source string) error {
+	modifiedLibraries := []*config.LibraryState{}
+	for _, library := range state.Libraries {
+		modifiedLibraries = append(modifiedLibraries, library)
+		modifiedAPIs := []config.API{}
+		for _, api := range library.APIs {
+			if api.ServiceConfig != "" {
+				// Do not change API if the service config has already been set.
+				modifiedAPIs = append(modifiedAPIs, api)
+				continue
+			}
+			apiPath := filepath.Join(source, api.Path)
+			serviceConfig, err := findServiceConfigIn(contentLoader, apiPath)
+			if err != nil {
+				return err
+			}
+			modifiedAPIs = append(modifiedAPIs, config.API{
+				Path:          api.Path,
+				ServiceConfig: serviceConfig,
+			})
+		}
+		modifiedLibraries[len(modifiedLibraries)-1].APIs = modifiedAPIs
+	}
+	state.Libraries = modifiedLibraries
+
+	return nil
+}
+
+func findServiceConfigIn(contentLoader func(file string) ([]byte, error), path string) (string, error) {
```

Comment:
Can you add comment to this method stating the heuristic it looks for? https://github.com/googleapis/librarian/issues/777#issuecomment-3044211865

Reply:
Done.